src/slow_hash_plug.c

```diff
@@ -147,90 +147,143 @@ void hash_process(union hash_state *state, const uint8_t *buf, size_t count)
 	keccak1600(buf, count, (uint8_t*)state);
 }
 
-void cn_fast_hash(const void *data, size_t length, char *hash)
+#if MBEDTLS_AESNI_HAVE_CODE == 2
+static inline void aesni_pseudo_encrypt_ecb(const uint8_t *exp_data, block * restrict c)
 {
-	union hash_state state;
-	hash_process(&state, data, length);
-	memcpy(hash, &state, HASH_SIZE);
+	__m128i cv = c->v;
+	const __m128i *dv = (const __m128i *)exp_data;
+	cv = _mm_aesenc_si128(cv, dv[0]);
+	cv = _mm_aesenc_si128(cv, dv[1]);
+	cv = _mm_aesenc_si128(cv, dv[2]);
+	cv = _mm_aesenc_si128(cv, dv[3]);
+	cv = _mm_aesenc_si128(cv, dv[4]);
+	cv = _mm_aesenc_si128(cv, dv[5]);
+	cv = _mm_aesenc_si128(cv, dv[6]);
+	cv = _mm_aesenc_si128(cv, dv[7]);
+	cv = _mm_aesenc_si128(cv, dv[8]);
+	cv = _mm_aesenc_si128(cv, dv[9]);
+	c->v = cv;
 }
+#endif
 
-void cn_slow_hash(const void *data, size_t length, char *hash)
+int cn_slow_hash(const void *data, size_t length, char *hash, void *memory)
 {
 #if MBEDTLS_AESNI_HAVE_CODE == 2
 	const int have_aesni = mbedtls_aesni_has_support(MBEDTLS_AESNI_AES);
 #endif
-	//uint8_t long_state[MEMORY]; // This is 2 MB, too large for stack
-	uint8_t *long_state = mem_alloc(MEMORY);
+	block *long_state = memory; // This is 2 MiB, too large for stack
+	OAES_CTX *aes_ctx = oaes_alloc();
 	union cn_slow_hash_state state;
-	uint8_t text[INIT_SIZE_BYTE];
-	uint8_t a[AES_BLOCK_SIZE] JTR_ALIGN(16);
-	uint8_t b[AES_BLOCK_SIZE];
-	uint8_t c[AES_BLOCK_SIZE] JTR_ALIGN(16);
-	uint8_t d[AES_BLOCK_SIZE];
+	block text[INIT_SIZE_BLK];
+	block a, b, c, d;
 	size_t i, j;
-	uint8_t aes_key[AES_KEY_SIZE];
-	OAES_CTX* aes_ctx;
 
 	hash_process(&state.hs, data, length);
```

Comment:
Shouldn't `aes_ctx` be checked here instead of at line 184? No need to call `hash_process` if the allocation fails.

Reply:
Either way works. Yes, we could skip `hash_process` if the allocation fails, but we really don't expect failure.

What I actually thought of is moving the `OAES_CTX *aes_ctx = oaes_alloc();` line to be just before the `if (!aes_ctx ||` line. Maybe that would be cleaner. I think it's primarily a matter of style.

Thank you for taking a look!

Reply:
> What I actually thought of is moving the `OAES_CTX *aes_ctx = oaes_alloc();` line to be just before the `if (!aes_ctx ||` line.

That's what I did in a subsequent commit.